javaslang/src/main/java/javaslang/collection/AbstractMultimap.java

```diff
@@ -175,10 +92,11 @@ public M distinctBy(Comparator<? super Tuple2<K, V>> comparator) {
         return (M) createFromEntries(iterator().distinctBy(comparator));
     }
 
+    @SuppressWarnings("unchecked")
     @Override
-    public <U> Multimap<K, V> distinctBy(Function<? super Tuple2<K, V>, ? extends U> keyExtractor) {
+    public <U> M distinctBy(Function<? super Tuple2<K, V>, ? extends U> keyExtractor) {
         Objects.requireNonNull(keyExtractor, "keyExtractor is null");
-        return createFromEntries(iterator().distinctBy(keyExtractor));
+        return (M) createFromEntries(iterator().distinctBy(keyExtractor));
```

Comment:
I think that if you change `createFromEntries` to `M` also, you won't nee these casts (multiple ones in this file)